README.md
# 🐹 API in GO to create Cloud Function

> An API for deploying Google Cloud Functions in Go.

![preview](./images/preview.png)

## 🎯 Features

- Create, update or delete a Cloud Function;
- Get a Cloud Function via function ID;
- Set IAM Policy to enable unauthenticated users invoke the Cloud Function;

## ⚙️ Create a private key to access the GCP

For this method, you'll need to [create a service account](https://cloud.google.com/docs/authentication/getting-started), and download a key.

1. In the GCP Console, go to the [Create service account key](https://console.cloud.google.com/apis/credentials/serviceaccountkey?_ga=2.44822625.-475179053.1491320180) page;
2. From the Service account drop-down list, select New service account;
3. In the Service account name field, enter a name;
4. From the Role drop-down list, select Project > Owner or if you wanna more restrict roles you can choose: `cloudfunctions.admin` (Cloud Functions Admin) and `iam.serviceAccountUser` (Service Account User);
   1. The service account needs to Cloud Functions Admin to have permission to update the IAM policies;
5. Click Create. A JSON file that contains your key downloads to your computer;
6. Save the JSON file to a safe place on your local machine (I create a hidden folder and save it in `~/.google-cloud-keys/key.json` );
7. Update the constant `projectID` with the name of your project at `main.go`;

## Run

```bash
GOOGLE_APPLICATION_CREDENTIALS=~/.google-cloud-keys/key.json go run *.go
```

Send the below request to http://localhost:8080/

## API

<details>
  <summary>Create or update a Cloud Function</summary>

#### Accepts HTTP `POST`

#### Path `/`

#### Request

```json
{
   "name":"cloud-function-test",
   "files":[
      {
         "name":"index.js",
         "content":"exports.helloWorld = (req, res) => {\n  let message = req.query.message || req.body.message || 'Hello World! 4';\n  res.status(200).send(message);\n};"
      },
      {
         "name":"package.json",
         "content":"{\n  \"name\": \"sample-http\",\n  \"version\": \"0.0.1\"\n}"
      }
   ]
}
```

#### Response

```json
{
   "message":"Created the Cloud Function"
}
```

#### Result

![preview](./images/result.png)

![preview](./images/Functions_–_Cloud_Functions_–_Google_Cloud_Platform.png)
</details>
<details>
  <summary>Get a Cloud Function via function ID</summary>

#### Accepts HTTP `GET`

#### Path `/`

#### Request

- `function_id` - string representing the cloud function ID

```bash
curl -X GET http://localhost:8080?function_id=trud_test
```

#### Response

```json
{
   "cloudFunction":{
      "availableMemoryMb":128,
      "entryPoint":"helloWorld",
      "httpsTrigger":{
         "url":"https://us-central1-vendasta-hackathon.cloudfunctions.net/trud_test"
      },
      "ingressSettings":"ALLOW_ALL",
      "name":"projects/vendasta-hackathon/locations/us-central1/functions/trud_test",
      "runtime":"nodejs8",
      "serviceAccountEmail":"vendasta-hackathon@appspot.gserviceaccount.com",
      "sourceUploadUrl":"https://storage.googleapis.com/gcf-upload-us-central1-ed9343f1-111c-435e-b16c-28bb09e2f13f/f3679a71-e48e-4c94-8d26-307775296a73.zip?GoogleAccessId=service-481416019804@gcf-admin-robot.iam.gserviceaccount.com&Expires=1590974019&Signature=HDKLbM4x1StRLa8ICqR%2B2R5WEHRmB%2BeqdM2e7btRh1Jb5%2BDqn5SMh1SziOh48ZYDDTjSHpxtQO17%2F3ZeopBXhtEEg2ytGF%2BzJXRA5C2k5BuTX4ULK9OyLQwhE1TDGK5DqJ4JZ%2Fnylfnpdvs4UtqpP4s3Rt4yBszBM5HfIwXYsK7S2HXmskEMc7U9a8rXP6QdqSEwRyTth%2FWTorLt8S9qB0VB8lX8l7xpIsesL0R0mxhfLhn63TQggkx61yOu9qa%2FShjcnqRquOGDEG%2BQ2zEsvKijCaGDZDLq%2Fx0qSvfXSHAQ41uHHnv%2B7Xy6R31j%2Bhe53H3lKLmvUrgL6oM%2Bc5maHQ%3D%3D",
      "status":"ACTIVE",
      "timeout":"60s",
      "updateTime":"2020-06-01T00:44:07.102Z",
      "versionId":"3"
   }
}
```

</details>
<details>
  <summary>Delete a Cloud Function via function ID</summary>

#### Accepts HTTP `DELETE`

#### Path `/`

#### Request

- `function_id` - string representing the cloud function ID

```bash
curl -X DELETE http://localhost:8080?function_id=xdne_xpto
```

#### Response

```json
{
   "cloudFunction":{
      "metadata":{
         "@type":"type.googleapis.com/google.cloud.functions.v1.OperationMetadataV1",
         "target":"projects/vendasta-hackathon/locations/us-central1/functions/xdne_xpto",
         "type":"DELETE_FUNCTION",
         "request":{
            "@type":"type.googleapis.com/google.protobuf.Empty"
         },
         "updateTime":"2020-07-11T00:15:13Z"
      },
      "name":"operations/dmVuZGFzdGEtaGFja2F0aG9uL3VzLWNlbnRyYWwxL3hkbmVfeHB0by9FUzlUX1N1RzVPaw"
   }
}
```

</details>
<details>
  <summary>Set IAM Policy to `allUsers` have access to invoker the Cloud Function</summary>

#### Accepts HTTP `POST`

#### Path `/policy`

#### Request

```json
{
   "function_id":"trud_test"
}
```

#### Response

```json
{
   "policy":{
      "bindings":[
         {
            "members":[
               "allUsers"
            ],
            "role":"roles/cloudfunctions.invoker"
         }
      ],
      "etag":"BwWm/F4UOXk=",
      "version":1
   }
}
```
</details>

## Libraries

- https://github.com/gofiber/fiber
- https://github.com/awslabs/git-secrets

### Upgrade Go Fiber

```bash
go get -u github.com/gofiber/fiber/v2
go get -u github.com/gofiber/fiber/v2/middleware/cors
go get -u github.com/gofiber/fiber/v2/middleware/logger
```

## 📚 References

- Cloud Function API: https://cloud.google.com/functions/docs/reference/rest
- Client Libraries: https://cloud.google.com/apis/docs/client-libraries-explained
- An API and CLI for deploying Google Cloud Functions in Node.js: https://github.com/JustinBeckwith/gcx
- Deploying from the Google Cloud Functions API: https://cloud.google.com/functions/docs/deploying/api
- Auto-generated Google APIs for Go: https://github.com/googleapis/google-api-go-client
- gcloud functions deploy: https://cloud.google.com/sdk/gcloud/reference/functions/deploy#--runtime
- IAM Policies to Cloud Function: https://cloud.google.com/functions/docs/reference/iam/roles
- Help keep your Google Cloud service account keys safe: taking charge of your security: https://cloud.google.com/blog/products/gcp/help-keep-your-google-cloud-service-account-keys-safe

## Author

- Cândido Sales - [@candidosales](https://twitter.com/candidosales)

## ⚠️ Copyright and license

Code and documentation copyright 2020-2030 the [Authors](https://github.com/candidosales/cloud-function-go/graphs/contributors) and Code released under the [MIT License](https://github.com/candidosales/cloud-function-go/blob/master/LICENSE). Docs released under [Creative Commons](https://creativecommons.org/licenses/by/3.0/).

CHANGELOG.md
# Changelog

## [1.3.0] - 2021-05-27

### Changed

- Upgrade GoFiber to 2.10.0
- Update Go Version `go mod edit -go=1.15`
- Update README

## [1.2.0] - 2020-10-02

### Changed

- Upgrade GoFiber to 2.0.5
- Update the README

## [1.1.1] - 2020-07-21

### Changed

- Upgrade GoFiber to 1.13.1
- Change to Node 8 to Node 10. Node 8 is deprecated: https://firebase.google.com/support/faq?authuser=0#expandable-10

## [1.1.0] - 2020-07-10

### Added

- Add delete method

## [1.0.1] - 2020-07-06

### Changed

- Update the README and improve the documentation
- Replace the JSON file to environment variable
- Update the image result.png

### Fixed

- Fix the endpoint to get Cloud Function

## [1.0.0] - 2020-07-05

### Added

- Add gitignore

### Changed

- Upgrade GoFiber


routes.go
package main

import (
	"fmt"

	"github.com/gofiber/fiber/v2"
)

func (app *AppConfig) setupRoutes() {
	app.fiber.Get("/ping", app.ping)
	app.fiber.Post("/", app.createCF)
	app.fiber.Get("/", app.getCF)
	app.fiber.Delete("/", app.deleteCF)
	app.fiber.Post("/policy", app.setPolicy)
}

func (app *AppConfig) ping(c *fiber.Ctx) error {
	return c.JSON(fiber.Map{"pong": "ok"})
}

func (app *AppConfig) createCF(c *fiber.Ctx) error {
	createCFRequest := &CreateCFRequest{}

	if err := c.BodyParser(&createCFRequest); err != nil {
		fmt.Printf("err[%#v] \n", err)
	}

	result, err := app.createCloudFunction(createCFRequest)
	if err != nil {
		return c.Status(503).Send([]byte(err.Error()))
	}

	return c.JSON(fiber.Map{"message": result})
}

func (app *AppConfig) getCF(c *fiber.Ctx) error {
	functionID := c.Query("function_id")

	result, err := app.getCloudFunction(functionID)
	if err != nil {
		return c.Status(400).Send([]byte(err.Error()))
	}

	return c.JSON(fiber.Map{"cloudFunction": result})
}

func (app *AppConfig) deleteCF(c *fiber.Ctx) error {
	functionID := c.Query("function_id")

	result, err := app.deleteCloudFunction(functionID)
	if err != nil {
		return c.Status(400).Send([]byte(err.Error()))
	}

	return c.JSON(fiber.Map{"cloudFunction": result})
}

func (app *AppConfig) setPolicy(c *fiber.Ctx) error {
	getCFRequest := &GetCFRequest{}

	if err := c.BodyParser(&getCFRequest); err != nil {
		fmt.Printf("err[%#v] \n", err)
	}

	result, err := app.setIAMPolicy(getCFRequest.FunctionID)
	if err != nil {
		return c.Status(400).Send([]byte(err.Error()))
	}

	return c.JSON(fiber.Map{"policy": result})
}


main.go
package main

import (
	"archive/zip"
	"bytes"
	"context"
	"errors"
	"fmt"
	"log"
	"net/http"
	"os"
	"strings"
	"time"

	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/fiber/v2/middleware/cors"
	"github.com/gofiber/fiber/v2/middleware/logger"
	"google.golang.org/api/cloudfunctions/v1"
	"google.golang.org/api/option"
)

const (
	projectID = "vendasta-hackathon"
	region    = "us-central1"
)

var (
	ErrFunctionIDIsEmpty        = errors.New("function_id is empty")
	ErrGoogleCredentialsIsEmpty = errors.New("GOOGLE_APPLICATION_CREDENTIALS variable is empty")
	ErrProjectIDIsInvalid       = errors.New("projectID constant is invalid. Please set the projectID")
)

func main() {
	ctx := context.Background()

	credentials := os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")
	if credentials == "" {
		log.Fatal(ErrGoogleCredentialsIsEmpty)
	}

	if projectID == "YOUR-PROJECT" {
		log.Fatal(ErrProjectIDIsInvalid)
	}

	cfService, err := cloudfunctions.NewService(ctx, option.WithCredentialsFile(credentials))
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("Cloud Function Service started ... ")

	appFiber := fiber.New()
	appFiber.Use(logger.New())
	appFiber.Use(cors.New())

	appConfig := AppConfig{
		projectID:      projectID,
		region:         region,
		cloudfunctions: cfService,
		fiber:          appFiber,
	}

	port := os.Getenv("PORT")
	if port == "" {
		port = ":8080"
	}

	appConfig.setupRoutes()
	err = appConfig.fiber.Listen(port)
	if err != nil {
		log.Fatal(err)
	}
}

func (app *AppConfig) createCloudFunction(createCFRequest *CreateCFRequest) (string, error) {
	start := time.Now()

	cfName := fmt.Sprintf("%s", strings.ToLower(createCFRequest.Name))
	name := app.getLocation() + "/functions/" + cfName

	fmt.Println("🤖 Deployment initialized.")
	fmt.Println("Packaging sources...")

	sourceUploadUrl := app.generateUploadURL()
	bodyBuffer := app.packFiles(createCFRequest.Files)

	fmt.Println("Uploading source...")

	_, err := app.uploadFileRequest(sourceUploadUrl, bodyBuffer)
	if err != nil {
		fmt.Printf("uploadFileRequest - err[%#v] \n", err)
		return "", err
	}
	requestBody := app.buildRequestBody(sourceUploadUrl, name)

	fmt.Println("Deploying function...")

	if app.cloudFunctionExists(name) {
		operation, err := app.cloudfunctions.Projects.Locations.Functions.Patch(name, requestBody).Do()
		if err != nil {
			fmt.Printf("Patch - err[%#v] \n", err)
			return "", err
		}

		if operation != nil {
			fmt.Printf("🚀 Update - Function deployed in %d seconds. \n", time.Now().Sub(start))
			return "Updated the Cloud Function", nil
		}

	} else {
		operation, err := app.cloudfunctions.Projects.Locations.Functions.Create(app.getLocation(), requestBody).Do()
		if err != nil {
			fmt.Printf("Create - err[%#v] \n", err)
		}

		if operation != nil {
			fmt.Printf("🚀 Create - Function deployed in %d seconds. \n", time.Now().Sub(start))
			return "Created the Cloud Function", nil
		}
	}
	return "", nil
}

func (app *AppConfig) getLocation() string {
	return fmt.Sprintf("projects/%s/locations/%s", app.projectID, app.region)
}

func (app *AppConfig) generateUploadURL() string {
	response, err := app.cloudfunctions.Projects.Locations.Functions.GenerateUploadUrl(app.getLocation(), &cloudfunctions.GenerateUploadUrlRequest{}).Do()
	if err != nil {
		fmt.Printf("err: %#v \n", err)
		return ""
	}

	return response.UploadUrl
}

func (app *AppConfig) packFiles(filesRequest []FilesRequest) *bytes.Buffer {
	// Create a buffer to write our archive to.
	buf := new(bytes.Buffer)

	// Create a new zip archive.
	w := zip.NewWriter(buf)
	defer w.Close()

	for _, file := range filesRequest {
		f, err := w.Create(file.Name)
		if err != nil {
			log.Fatal(err)
		}
		_, err = f.Write([]byte(file.Content))
		if err != nil {
			log.Fatal(err)
		}
	}

	// Make sure to check the error on Close.
	err := w.Close()
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("📦 Source code packaged.")
	return buf
}

func (app *AppConfig) uploadFileRequest(sourceUploadURL string, body *bytes.Buffer) (*http.Response, error) {

	req, err := http.NewRequest(http.MethodPut, sourceUploadURL, body)
	req.Header.Set("Content-Type", "application/zip")
	req.Header.Set("X-Goog-Content-Length-Range", "0,104857600") // https://cloud.google.com/storage/docs/json_api/v1/parameters#xgoogcontentlengthrange

	client := &http.Client{}
	resp, err := client.Do(req)

	if err != nil {
		return nil, err
	} else {
		body := &bytes.Buffer{}
		_, err := body.ReadFrom(resp.Body)
		if err != nil {
			return nil, err
		}

		err = resp.Body.Close()
		if err != nil {
			return nil, err
		}

		if resp.StatusCode == 200 {
			fmt.Println("🛸 Source uploaded to cloud.")
		}
	}

	return resp, err
}

func (app *AppConfig) buildRequestBody(sourceUploadURL string, name string) *cloudfunctions.CloudFunction {
	return &cloudfunctions.CloudFunction{
		Name:              name,                           // Required - Name must start with a letter followed by up to 62 letters, numbers or hyphens, and cannot end with a hyphen
		AvailableMemoryMb: 128,                            // Optional
		EntryPoint:        "helloWorld",                   // Required
		Runtime:           "nodejs10",                     // Required
		HttpsTrigger:      &cloudfunctions.HttpsTrigger{}, // Required
		SourceUploadUrl:   sourceUploadURL,                // Required
		IngressSettings:   "ALLOW_ALL",                    // Optional
	}
}

func (app *AppConfig) cloudFunctionExists(name string) bool {
	_, err := app.cloudfunctions.Projects.Locations.Functions.Get(name).Do()
	if err != nil {
		return false
	}
	return true
}

func (app *AppConfig) getCloudFunction(functionID string) (*cloudfunctions.CloudFunction, error) {
	if functionID == "" {
		return nil, ErrFunctionIDIsEmpty
	}

	name := app.getLocation() + "/functions/" + functionID

	response, err := app.cloudfunctions.Projects.Locations.Functions.Get(name).Do()
	if err != nil {
		return nil, err
	}
	return response, err
}

func (app *AppConfig) deleteCloudFunction(functionID string) (*cloudfunctions.Operation, error) {
	if functionID == "" {
		return nil, ErrFunctionIDIsEmpty
	}

	name := app.getLocation() + "/functions/" + functionID

	response, err := app.cloudfunctions.Projects.Locations.Functions.Delete(name).Do()
	if err != nil {
		return nil, err
	}
	return response, err
}

func (app *AppConfig) setIAMPolicy(functionID string) (*cloudfunctions.Policy, error) {

	if functionID == "" {
		return nil, ErrFunctionIDIsEmpty
	}

	name := app.getLocation() + "/functions/" + functionID

	setIAMPolicyRequest := &cloudfunctions.SetIamPolicyRequest{
		Policy: &cloudfunctions.Policy{
			Bindings: []*cloudfunctions.Binding{
				{
					Role:    "roles/cloudfunctions.invoker",
					Members: []string{"allUsers"},
				},
			},
		},
	}

	response, err := app.cloudfunctions.Projects.Locations.Functions.SetIamPolicy(name, setIAMPolicyRequest).Do()
	if err != nil {
		return nil, err
	}
	return response, err
}
